### ISSUE TITLE
Response cards are not shown with voice input

### ISSUE THREAD

**Issue by dsbyprateekg** [aws-samples/aws-lex-web-ui#213]
Hi,

I am facing a strange issue with AWS Lex web ui.
When I use voice input in Test bot console, it is not showing response cards as shown in below screenshot-
![no_respnse_voice](https://user-images.githubusercontent.com/30830541/83320571-8e8a1780-a266-11ea-9154-eb9dd166d5c5.JPG)

But when I use text input, it's working-
![response_text](https://user-images.githubusercontent.com/30830541/83320610-b6797b00-a266-11ea-8a43-aba54f913fc0.JPG)

When I verify the same in deployed web app, issue becomes worst. There bot is even not taking voice input although text input is working as expected.
![webapp_not_taking_voice](https://user-images.githubusercontent.com/30830541/83320671-356eb380-a267-11ea-8ea1-e511bff4e6f0.JPG)

Please help to resolve both issues.

**Comment by bobpskier** [comment]
I tested the URL referenced in your image to see if I could replicate the issues. I'm using chrome and as soon as I clicked the microphone, the browser asked me if I authorized enabling the microphone. I clicked on yes and was able to interact using audio with the bot. I suspect a browser setting is disabling access to the microphone. The browser controls this capability. Can you check to see if use of the microphone is enabled. 

I tested use of "hi" from both text and audio methods and observed the same behavior as you reported using the Lex test console. When using audio, I did not observe a response card being returned to the client after the request to Lex. Is this the desired behavior? I believe the flexibility to return a response card after audio interaction is available.

**Comment by dsbyprateekg** [comment]
With voice input, response card is not being returned.
This is not the expected behavior.
I have checked the same with order flowers example bot and there also with voice input response card was not being returned.
So if this option is available, can you please guide me is there any specific setting to enable this option? Please suggest.

**Comment by rstrahan** [comment]
Hi.. Response Cards are not natively supported by Lex in voice mode (postContent API). However, when using the postContent API call, response cards can be passed to lex-web-ui using session attributes in the appContext.responseCard key of the sessionAttributes object.

For details see: https://github.com/aws-samples/aws-lex-web-ui/blob/master/lex-web-ui/README.md

**Comment by dsbyprateekg** [comment]
@rstrahan this option also I had tried. I had set the response cards in session attributes and it was visible also in my session attribute variable but in web app voice input is not even working. I have allowed the microphone settings and tested in 2 different laptops but issue still remains.
I am using Python for lambda function, please see this snippet of setting response cards in session-
![cards_in_response](https://user-images.githubusercontent.com/30830541/83343545-84752100-a319-11ea-9906-6974f8bf4ad4.JPG)

**Comment by dsbyprateekg** [comment]
I was facing voice related issues in web app because it seems voice sdk is updated in browsers.
Now for voice input click one time on the voice icon and speak. No need to click again on voice icon. Click only if you want to stop the flow.

Although this sudden change in voice is very poor. Many input voice words are not recognized correctly by the browser which were working earlier. It seems they have removed noise cancellation.

And @rstrahan I am able to show response cards by setting in session attributes as you suggested. I was doing wrong previously. Thanks for your suggestion. My issue is resolved now.

**Comment by bhaveshkumar80** [comment]
Hii @dsbyprateekg, How you have solved this issue. I have added response card in session attribute also but not working for me.

**Comment by jjayeshpawar** [comment]
@rstrahan @dsbyprateekg I tries using *'sessionAttributes' : {'appContext': json.dumps({'responseCard': responseCard})}* in response but still response card is not showing when I try to use voice input. Kindly help. @dsbyprateekg can you please share us what worked for you? Thanks

**Comment by karanakatle** [comment]
@rstrahan @dsbyprateekg can you please let me know what works for you while adding appcontext in session attributes
I am trying to do the same but not getting results?

**Comment by cansado2930** [comment]
@rstrahan @dsbyprateekg Could you explain with more details how to fix the problem with card and voice? We have the same problem, the user writes and it works correctly, but if the user speak, it said "there was an error" and it doesn't show nothing. 
Thank you!

**Comment by rstrahan** [comment]
Hi @karanakatle @cansado2930 et all,
If it helps you can reverse engineer a working example using [QnABot](https://www.amazon.com/qnabot) - you can use the built-in lex web UI client for testing.. Import the blog-samples-final package.. This has an example question with an image/button response card for the question "What is the Echo Show".  You can try the question from the web client with either text or voice.. in both cases it renders the response card.  The lambda fullfillment function logs will show the final response object send by the fulfillment fn back to Lex.. like this:
```
2022-03-10T17:44:57.777Z	a833918e-a637-470a-91e6-7e708d68e3a5	INFO	final: 
{
	"sessionState": {
		"sessionAttributes": {
			"localTimeZone": "America/New_York",
			"qnabotcontext": "{\"userPreferredLocale\":\"en\",\"userLocale\":\"en\",\"previous\":{\"qid\":\"Alexa.001\",\"q\":\"what is the echo show\"},\"navigation\":{\"next\":\"\",\"previous\":[],\"hasParent\":false}}",
			"userDetectedLocale": "en",
			"userDetectedLocaleConfidence": "0.9982491731643677",
			"topic": "EchoShow",
			"appContext": "{\"altMessages\":{\"markdown\":\"\",\"ssml\":\"\"},\"responseCard\":{\"version\":\"1\",\"contentType\":\"application/vnd.amazonaws.card.generic\",\"genericAttachments\":[{\"title\":\"Echo Show\",\"imageUrl\":\"https://images-na.ssl-images-amazon.com/images/I/61OddH8ddDL._SL1000_.jpg\",\"buttons\":[{\"text\":\"QnABot\",\"value\":\"What is Q and A bot?\"},{\"text\":\"FireTV\",\"value\":\"What is Amazon Fire TV?\"}]}]}}",
			"qnabot_qid": "Alexa.001",
			"qnabot_gotanswer": "true",
			"connect_nextPrompt": ""
		},
		"dialogAction": {
			"type": "Close"
		},
		"intent": {
			"name": "QnaIntent",
			"state": "Fulfilled"
		}
	},
	"messages": [{
			"contentType": "PlainText",
			"content": "Echo Show brings you everything you love about Alexa, and now she can show you things. She is the perfect companion for Q and A Bot."
		},
		{
			"contentType": "ImageResponseCard",
			"imageResponseCard": {
				"title": "Echo Show",
				"subTitle": "",
				"imageUrl": "https://images-na.ssl-images-amazon.com/images/I/61OddH8ddDL._SL1000_.jpg",
				"buttons": [{
						"text": "QnABot",
						"value": "What is Q and A bot?"
					},
					{
						"text": "FireTV",
						"value": "What is Amazon Fire TV?"
					}
				]
			}
		}
	]
}

```
![image](https://user-images.githubusercontent.com/10953374/157726857-96fb7ad7-9cd4-4107-ae93-7d1ea3aa8f08.png)

**Comment by bobpskier** [comment]
@cansado2930 @karanakatle @rstrahan 

LexWebUi does not itself take SSML from the altMessages session attribute and generate audio from the response. Audio generation is the responsibility of the Lex service, or the Alexa service, or the Connect service. The lambda handler function must return valid SSML as part of the messages response payload in Lex V2 in order to generate audio via the Lex service. 

LexWebUi will add markdown or html from the appContext.altMessages session attribute as part of the message output if present.

Audio output will only be generated by Lex service as a response to the postContent (v1) or recognizeUtterance (v2) invocations for LexWebUi. 

ResponseCard is not translated to audio by the Lex Service. If you need to display a response card and present the same content via audio presentation you must include this as SSML content in the return from your lambda function. The following is an example for Lex V2 that does just this returned from a lambda handler. Lex V1 response is entirely different. 

```
{
  "sessionState": {
    "sessionAttributes": {   
      "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:97.0) Gecko/20100101 Firefox/97.0",
      "appContext": "{\"altMessages\":{\"ssml\":\"<speak>Do you want to specify each fuel rate input as kilograms per second or pounds per second? Say 1 for kilograms or 2 for pounds.</speak>\",\"markdown\":\"Do you want to specify each **fuel rate input** as lbs per second or kg per second?\"},\"responseCard\":{\"version\":\"1\",\"contentType\":\"application/vnd.amazonaws.card.generic\",\"genericAttachments\":[{\"title\":\"Options\",\"buttons\":[{\"text\":\"kg\",\"value\":\"1\"},{\"text\":\"LBS\",\"value\":\"2\"}]}]}}",
      "undefined": "{}"
    },
    "dialogAction": {
      "type": "Close"
    },
    "intent": {
      "name": "QnaIntent",
      "state": "Fulfilled"
    }
  },
  "messages": [
    {
      "contentType": "SSML",
      "content": "<speak>Do you want to specify each fuel rate input as kilograms per second or pounds per second? Say 1 for kilograms or 2 for pounds.</speak>"
    },
    {
      "contentType": "ImageResponseCard",
      "imageResponseCard": {
        "title": "Options",
        "buttons": [
          {
            "text": "kg",
            "value": "1"
          },
          {
            "text": "LBS",
            "value": "2"
          }
        ]
      }
    }
  ]
}
```
The audio response from the lex service will be generated from the messages array element based on the contentType SSML. Notice the message literally contains prompts in the speak section that also are included in the response card. 

The next most important element is the sessionState.sessionAttributes.appContext.altMessages.markdown. This contains content that will be rendered by LexWebUi as markdown. It must be specified as markdown. A separate html property is available. 

The next most important elements are the sessionState.sessionAttributes.appContext.altMessages.responseCard  element and the second message in the messages array which is also a ResponseCard. I believe the content from the altMessages.responseCard will override that in the second element in the messages array. One of the primary drivers behind the use of altMessages.responseCard was to increase the number of buttons that can be displayed as Lex places restrictions on the number of buttons that can be provided in a response. 

This is a V2 example. Unfortunately the V1 response format is different but the same concepts abut the use of appContext.allMessages.markdown, appContext.allMessages.responseCard, and appContext.allMessages.html remain the same. 

To summarize, the complete audio response including any buttons that you might have in a responseCard must also be provided in the SSML content. Use appContext.altMessages.markdown to specify any text to render in the message bubble.

**Comment by karanakatle** [comment]
Hey thank you @bobpskier for the help, issues are resolved now.

**Comment by cansado2930** [comment]
Thank you @rstrahan @bobpskier by your help. We have been able to fix the problem.
